Enable symbolication and source context for net9.0-android

jamescrosswell opened the pull request:
Closes https://github.com/getsentry/sentry-dotnet/issues/3761:
- https://github.com/getsentry/sentry-dotnet/issues/3761

Fixes https://github.com/getsentry/sentry-dotnet/issues/3983: 
- https://github.com/getsentry/sentry-dotnet/issues/3983

Fixes https://github.com/getsentry/sentry-dotnet/issues/3985:
- https://github.com/getsentry/sentry-dotnet/issues/3985

## Summary

.NET 9 does introduced a new AssemblyStore format for APK files. Reading the new format is demonstrated in [assembly-store-reader-mk2](https://github.com/dotnet/android/blob/5ebcb1dd1503648391e3c0548200495f634d90c6/tools/assembly-store-reader-mk2/Main.cs). This PR adds support for reading the new format to the `AndroidAssemblyReaderFactory`. 

Also, the reason we didn't pick this up when we moved to `net9.0` is that the device tests are/were only running against `net8.0` targets. In this PR we'll try to run the device tests on `net8.0-android/ios` and `net9.0-android/ios`. 

## Notes for Reviewers

There's a lot of code here. However most of it is simply copied/vendored in from two external repositories.

### ELFSharp

**For the purposes of review, you can ignore _all_ of the files in the "src/Sentry.Android.AssemblyReader/ELFSharp" folder.** 

The V2 format for APKs stores all assemblies in in an RID specific folder but Android requires all files in that folder to be ELF files... so during the build process any exe, dll, pdb or config files stored in there get "packed" into ELF files... we need to use ElfSharp to unpack these. 

To avoid adding any dependencies, the code for ELFSharp has been copied in to the Sentry.Android.AssemblyReader project (with appropriate attribution). 

### dotnet/android

**For the purposes of review, you can ignore _most_ of the files in the "src/Sentry.Android.AssemblyReader/V2" folder.** 

`src/Sentry.Android.AssemblyReader/V2/AndroidAssemblyStoreReaderV2.cs` and `AndroidAssemblyDirectoryReaderV2` are classes I wrote (although `AndroidAssemblyDirectoryReaderV2.ArchiveAssemblyHelper` was vendored in). Otherwise, all of the files in `src/Sentry.Android.AssemblyReader/V2` have been copied from the `dotnet/android` repo, with appropriate attribution at the top of each file.

No functional changes were made to any of that code. The only alterations were:
- Visibility changed from `public `to `internal` 
- Unused code from files that we vendored in has been removed
- I removed the dependency on `Xamarin.LibZipSharp` (using the built in [ZipFile](https://learn.microsoft.com/en-us/dotnet/api/System.IO.Compression.ZipFile?view=net-8.0) class instead)
- A little bit of `Ctrl + .` in resharper to get the code to conform to our coding conventions

## References

Relates to:
- https://github.com/getsentry/sentry-dotnet/pull/2050
- https://github.com/dotnet/android/pull/8478
- https://elfsharp.it/

jamescrosswell commented:
@vaind wondering if you'd be able to look at the Notes for Reviewers on this PR. 

In particular, wondering if/how you dealt with ELF files in the first version of the AndroidAssemblyReader?

vaind commented:
With all the vendored-in code this is pretty hard to review. Can you link a diff showing your changes to the vendored code?

jamescrosswell commented:
> With all the vendored-in code this is pretty hard to review. Can you link a diff showing your changes to the vendored code?

I can't think of any simple way to do that. [This earlier commit](https://github.com/getsentry/sentry-dotnet/pull/4033/commits/339f4ef2bcec8b2710a29e50b5e2b921806eb6d4) represents the minimal work to get most of the vendored in code copied into our project without breaking the build... before any of our code actually uses that new code. So you could diff individual vendored in files with their equivalent in that commit. At that stage I was only trying to get the AssemblyStoreExplorer working though (so none of the code required to get `src/Sentry.Android.AssemblyReader/V2/AndroidAssemblyDirectoryReaderV2.cs` working). 

I've added some details to the PR description for reviewers, that hopefully makes things a little easier. 

For the most part, I think you can ignore the vendored (treat it kind of like a 3rd party library) as well as any changes I made to that code as my changes weren't functional in nature (I just removed stuff and tidied it up a little). The only exception **_might_** be the [change I made to remove `Xamarin.LibZipSharp`](https://github.com/getsentry/sentry-dotnet/pull/4033/commits/fc3e17a2b64b1e55d1f8e3a81f6254c82cdfea16)... potentially that's related to the [1 unit test that is still failing](https://github.com/getsentry/sentry-dotnet/blob/45769f95be5f4d7fd5b77acdb00ffa5e782214cb/test/Sentry.Android.AssemblyReader.Tests/AndroidAssemblyReaderTests.cs#L87)... although I don't immediately see how. [This one](https://github.com/getsentry/sentry-dotnet/blob/45769f95be5f4d7fd5b77acdb00ffa5e782214cb/test/Sentry.Android.AssemblyReader.Tests/AndroidAssemblyReaderTests.cs#L85) passes, after all. It's odd that we can read `Mono.Android.dll` from an archive that doesn't use the assembly store but we can't do the same for `System.Runtime.dll`.

bruno-garcia commented on scripts/device-test.ps1:
```diff
@@ -21,13 +22,16 @@ $CI = Test-Path env:CI
-    $tfm = 'net8.0-'
+    if (!$Tfm)
+    {
+        $Tfm = 'net8.0'
```
once we clear all callsites to pass a tfm, we should delete this and throw an error instead.

jamescrosswell replied:
This is just a default argument, if you don't specify a target framework. We could try to change the default to `latest` or something like that - would need a little bit of extra scripting to work out what latest was.

bruno-garcia commented on src/Sentry.Android.AssemblyReader/ELFSharp/MachO/Dylib.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace ELFSharp.MachO
+{
+    internal abstract class Dylib : Command
```
do we need this? this is a mac thing

jamescrosswell replied:
Possibly not - I didn't make any attempt to trim off unused code from ELFSharp (it's had the minimal changes required to get our project to build with it). 

We could try to shrink it down. The linker probably removes unused code paths when publishing on Android but I guess we could slightly improve linker times by removing the code altogether.

Worth it?

bruno-garcia commented on src/Sentry.Android.AssemblyReader/ELFSharp/MachO/FatArchiveReader.cs:
```diff
@@ -0,0 +1,32 @@
+using System.IO;
+using ELFSharp.Utilities;
+
+namespace ELFSharp.MachO
```
Do we need antyhing from `.MachO` namespace?  not relevant for Android

bruno-garcia reviewed (commented):
I just skimmed this change. Ideally someone can do a proper review
